### ISSUE TITLE
Switch from npm to yarn to allow cross-platform sharing of lockfile

### ISSUE THREAD

**Issue by nevik** [mozilla/voice-web#477]
I had noticed a while back that on my Mac, doing an `npm install` (note: npm v5 and later) always modifies the `package-lock.json` that is committed to the repo.

When I looked into this more today, I found that this is an issue with how npm handles optional dependencies (in our case, that's [`fsevents`](https://www.npmjs.com/package/fsevents), a macOS-only dependency of things that watch filesystem changes):

Since this is an optional dependency which can only be installed on macOS, npm will install it and add it to package-lock.json whenever `npm install` is used on a Mac, but will remove it (if present) from package-lock.json whenever `npm install` is run on another platform.

From what I could find, there are currently at least two issues in the npm repo describing/discussing this: npm/npm#17722 and npm/npm#18135.

Since a fix for this doesn't seem to be in sight, and [there's a report that this isn't an issue with yarn](https://github.com/npm/npm/issues/17722#issuecomment-332450218), I suggest that we switch to [yarn](https://yarnpkg.com/) for package management.

@mikehenrty, thoughts?

**Comment by mikehenrty** [comment]
I'm fine with switching to yarn (or can we support both)?

Either way, as long as it's not a big change switching back and forth, I think addressing the workflow nit you mentioned would be good (since we have mac, windows, and linux developers).

**Comment by nevik** [comment]
The idea behind using a lockfile (be it yarn's or npm's) is to ensure that each installation uses the exact same versions of all dependencies, i.e. reproducibility. (Doing it any other way usually leads to weird problems due to inconsistent deps sooner or later, especially with a CI system involved.)

You can technically always use yarn or npm to install locally (no matter which we "officially" use), and you can add dependencies to package.json with either one.
However, when not using the "official" one, you run the risk of the aforementioned inconsistency issues (and when adding dependencies, you're not updating the lockfile that most others use).

To summarize/bulletize:
* switching is easy (delete node_modules, re-install using yarn, remove package-lock.json and commit yarn.lock file)
* does not prevent people from using npm locally, though it has some gotchas when contributing back
* if we switch, we should update the travis build to also use yarn

**Comment by mikehenrty** [comment]
@nevik thanks for the background here. again, i'm fine with officially switching but we would need to add some documentation to explain installing yarn and why we chose it (over the default npm).

two other, lower footprint fixes would be:
1. removing package-lock.json from the repo
2. turn off optional dependency installation in .npmrc

Of the two above, I actually like 2. That way, development servers would run the same on everyone's machine, which could cut down on bugs. what do you think?

**Comment by nevik** [comment]
@mikehenrty, option 2 is discussed in npm/npm#18135, the specific issue why I don't like that approach is:

> Sometimes optional doesn't actually mean optional, but instead means _may be operating system specific_.

In our case (`fsevents` being the package in question, for now), that _might_ work out fine, but I've seen reports in those two issue threads I linked that lack of `fsevents` on macOS can actually case very high CPU and memory usage.

Option 1 I would avoid, esp. since we're aiming to introduce automated tests, because I've had bad experiences with non-locked dependencies and CI builds in the past. It can be quite difficult to track down that the reason why a specific test works locally but fails on CI is that one dependency is at version 1.7.5 locally but 1.7.4 on CI (since not all open source projects practice sem-ver reliably).

Since we now have an updated REAME and contribution guidelines, I think switching to yarn isn't actually such a big footprint.

I'll open a PR for the switch, then we'll see what it actually looks like.

**Comment by mikehenrty** [comment]
We use yarn now! \o/

https://github.com/mozilla/voice-web/commit/cdc936beb32c1f9cd97a3a041cd15ea3e22f2da7

**Comment by nevik** [comment]
🎉🎉🎉🎉
